python/pyarrow/serialization.pxi

```diff
@@ -0,0 +1,233 @@
+    type_id = serialized_obj["_pytype_"]
+
+    if "pickle" in serialized_obj:
+        # The object was pickled, so unpickle it.
+        obj = pickle.loads(serialized_obj["data"])
+    else:
+        assert type_id not in types_to_pickle
+        if type_id not in whitelisted_types:
+            raise "error"
+        type = whitelisted_types[type_id]
+        if type_id in custom_deserializers:
+            obj = custom_deserializers[type_id](serialized_obj["data"])
+        else:
+            # In this case, serialized_obj should just be the __dict__ field.
+            if "_pa_getnewargs_" in serialized_obj:
+                obj = type.__new__(type, *serialized_obj["_pa_getnewargs_"])
+            else:
+                obj = type.__new__(type)
+                serialized_obj.pop("_pytype_")
+                obj.__dict__.update(serialized_obj)
+    return obj
+
+pyarrow_serialize_callback = <PyObject*> serialization_callback
+
+pyarrow_deserialize_callback = <PyObject*> deserialization_callback
+
+# Main entry point for serialization
+def serialize_sequence(object value):
+    cdef int32_t recursion_depth = 0
+    cdef PythonObject result = PythonObject()
+    cdef c_vector[PyObject*] sequences
+    cdef shared_ptr[CArray] array
+    cdef c_vector[PyObject*] tensors
+    cdef PyObject* tensor
+    cdef shared_ptr[CTensor] out
+    sequences.push_back(<PyObject*> value)
+    check_status(SerializeSequences(sequences, recursion_depth, &array, tensors))
+    result.batch = MakeBatch(array)
+    num_tensors = 0
+    for tensor in tensors:
+        check_status(NdarrayToTensor(c_default_memory_pool(), <object> tensor, &out))
+        result.tensors.push_back(out)
+        num_tensors += 1
+    return result, num_tensors
+
+# Main entry point for deserialization
+def deserialize_sequence(PythonObject value, object base):
+    cdef PyObject* result
+    check_status(DeserializeList(deref(value.batch).column(0), 0, deref(value.batch).num_rows(), <PyObject*> base, value.tensors, &result))
+    return <object> result
+
+def write_python_object(PythonObject value, int32_t num_tensors, NativeFile sink):
+    cdef shared_ptr[OutputStream] stream
+    sink.write_handle(&stream)
+    cdef shared_ptr[CRecordBatchStreamWriter] writer
+    cdef shared_ptr[CSchema] schema = deref(value.batch).schema()
+    cdef shared_ptr[CRecordBatch] batch = value.batch
+    cdef shared_ptr[CTensor] tensor
+    cdef int32_t metadata_length
+    cdef int64_t body_length
+
+    with nogil:
+        # write number of tensors
+        check_status(stream.get().Write(<uint8_t*> &num_tensors, sizeof(int32_t)))
+
+        check_status(CRecordBatchStreamWriter.Open(stream.get(), schema, &writer))
+        check_status(deref(writer).WriteRecordBatch(deref(batch)))
+        check_status(deref(writer).Close())
+
+    for tensor in value.tensors:
+        check_status(WriteTensor(deref(tensor), stream.get(), &metadata_length, &body_length))
+
+def read_python_object(NativeFile source):
```

Comment:
I'm wondering a bit why `read_python_object` and `write_python_object` need to be public. If you pass a NativeFile to `serialize_*` then it should write the result to that, otherwise return a Buffer or byte string